src/rpc/message.cpp

```diff
@@ -211,82 +140,89 @@ cryptonote::rpc::error FullMessage::getError()
   return err;
 }
 
-FullMessage FullMessage::requestMessage(const std::string& request, Message* message)
+std::string FullMessage::getRequest(const std::string& request, const Message& message, const unsigned id)
 {
-  return FullMessage(request, message);
-}
+  rapidjson::StringBuffer buffer;
+  {
+    rapidjson::Writer<rapidjson::StringBuffer> dest{buffer};
 
-FullMessage FullMessage::requestMessage(const std::string& request, Message* message, rapidjson::Value& id)
-{
-  auto mes = requestMessage(request, message);
-  mes.setID(id);
-  return mes;
-}
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, jsonrpc, std::string{"2.0"});
```

Comment:
Since we are not moving the `std::string{"2.0"}` here (toJsonValue does not seem to have an overload taking an rvalue for `std::string`), would it make sense to have a `static const std::string` somewhere containing the RPC version number? It would make it easier to change it, should we ever need to in the future.

Reply:
Why is the std::string needed here at all? `INSERT_INTO_JSON_OBJECT(dest, jsonrpc, "2.0")` would work, as would a `static const char[]`.

Reply:
I don't remember why I specified the `std::string` explicitly. I think it was to remind myself to update the overload to a `boost::string_ref`. I will just do this now.

Reply:
I just realized that `string_ref` doesn't have an array-ref constructor overload (unlike, say, `rapidjson::StringRef`), only `const char *`. So, `strlen()` will be invoked on the argument every time one is constructed.

Obviously, that's very minor, but also trivial to avoid, using a single `static const boost::string_ref = "2.0"`.

Reply:
This is a situation where compiler might do the conversion automagically. The bigger issue is that string literals get trapped by `toJsonValue` for hex conversion. Which is why I did this - punt this to a future edit.

Reply:
@ndorf According to [boost.String_ref documentation](https://www.boost.org/doc/libs/1_72_0/libs/utility/doc/html/string_ref.html) the relevant constructor is declared as follows:

`BOOST_CONSTEXPR basic_string_ref(const charT* str);`

In our case, since we are using `c++11` and have NOT defined `BOOST_NO_CXX11_CONSTEXPR` the constructor will be constexpr qualified, so I don't think we'll have to worry about runtime calls to strlen.

Reply:
The documentation for `boost::string_ref` is incorrect. Only the overload with the length is `constexpr`. I tested `boost::string_view` (conforming to the new standard and not available in Boost 1.58) which has the relevant constructor marked as `constexpr`.  It is only usable as `constexpr` in C++17 or with a change in the default `traits` template argument. 

Disassembling revealed that Gcc and Clang will do runtime length computations in the usual call syntax. The programmer has to manually force `constexpr` behavior to get the compilers to compute the length at compile-time. A shame.

I somehow forgot that I already switched the `toJsonValue` overload to `boost::string_ref`, so I updated this portion of code with `boost::string_ref{"2.0", 3}` which is a bit more verbose but still easy.

Reply:
Sorry to bump this, but I've been away for a bit. A modest proposal:
```
template <size_t Len>
auto static_string_ref(const char (&str)[Len])
{
    return boost::string_ref(str, Len);
}
```

Reply:
You mean like [the instance I slipped into the codebase already](https://github.com/monero-project/monero/blob/d1cf16364fb1901c2940ce66024688ea3fe3fc48/contrib/epee/src/http_auth.cpp#L87)? I can find a header somewhere for it.